pkg/ring/replication_strategy.go

```diff
@@ -4,28 +4,43 @@ import (
 	"fmt"
 )
 
-// replicationStrategy decides, given the set of ingesters eligible for a key,
+type ReplicationStrategy interface {
+	// Filter out unhealthy instances and checks if there're enough instances
+	// for an operation to succeed. Returns an error if there are not enough
+	// instances.
+	Filter(cfg Config, instances []IngesterDesc, op Operation) (healthy []IngesterDesc, maxFailures int, err error)
+
+	// ExtendReplicaSet returns true if given an instance that's going to be
+	// added to the replica set, the replica set size should be extended by 1
+	// more instance for the given operation.
+	ExtendReplicaSet(instance IngesterDesc, op Operation) bool
```

Comment:
This method is not extending anything, and is a check instead. Can we name it such?

Reply:
It returns whether the replica set should be extended. Could you suggest a better name, please?

Reply:
`CanUseIngesterForOperation` or `IsOperationAllowedForInstance`.

Reply:
I thought about a similar naming, but it's not really what it does. We still keep the instance, but we do extend the replica set size.

Reply:
But not in this method. This method doesn't even take replica set as a parameter, nor does it operate on or extend one, so current name is confusing.

Reply:
I see your point, but I still believe something like `CanUseIngesterForOperation()` (or similar) is not correct either, because what this function tells is whether the ring should extend or not the replica set. Maybe `ShouldExtendReplicaSet()` sounds better?

Reply:
What about `InstanceAllowedInReplicaSet` ?

Reply:
No, that's still wrong...

Reply:
I think `ShouldExtendReplicaSet` describes this best. After looking into the code, it actually does something different than I thought.

Reply:
It's a weird thing to put into this interface  :-(

Reply:
> It's a weird thing to put into this interface :-(

Why? It's part of the replication strategy.

Reply:
> I think ShouldExtendReplicaSet describes this best.

Renamed in the meanwhile.

Reply:
> Renamed in the meanwhile.
Thanks.

> Why? It's part of the replication strategy.

You're probably right. I still find our replication strategy code quite confusing, but that's just me.

Thanks for addressing my feedback, feel free to merge.